curriculum/challenges/english/25-front-end-development/lab-hash-table/67ed03ac474c48692f41749e.md

```diff
@@ -0,0 +1,400 @@
+    sig = inspect.signature(ht.add)
+    len(sig.parameters) == 2
+  `))
+})
+```
+
+The `HashTable` class should have a `remove` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('remove')`))}})
+```
+
+The `remove` method should take a key as a parameter.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.remove)
+    len(sig.parameters) == 1
+  `))
+})
+```
+
+When you try to remove a key that does not exist in the collection, it should not raise an error or remove anything.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        ht.add("rose", "flower")
+        index = ht.hash("rose")
+        original = ht.collection.copy()
+
+        ht.remove("tulip")
+
+        assert ht.collection == original
+        assert "rose" in ht.collection[index]
+        assert "tulip" not in ht.collection.get(index, {})
+    `) 
+})
+```
+
+
+When the hash value is same, the `remove` method should only remove that key and not the entire hash table for that hash value.
+
+
+```js
+({
+  test: () => runPython(`
+    ht = HashTable()
+    ht.add("rose", "flower")
+    index = ht.hash("rose")
+
+    assert index in ht.collection
+
+    ht.remove("rose")
+
+    assert index not in ht.collection
+  `)
+})
+```
+
+
+The `HashTable` class should have a `lookup` method.
+
+```js
+({test: () => {assert(runPython(`_Node(_code).find_class('HashTable').has_function('lookup')`))}})
+```
+
+The `lookup` method should take a key as the parameter.
+
+```js
+({ test: () => assert(runPython(`
+    import inspect 
+    ht= HashTable()
+    sig = inspect.signature(ht.lookup)
+    len(sig.parameters) == 1
+  `))
+})
+```
+
+`HashTable.hash("golf")` should return `424`.
+
+```js
+({ 
+    test: () => runPython(`
+        ht = HashTable()
+        hash_result = ht.hash("golf")
+        assert hash_result == 424
+
+        # prevent hardcoding
+        assert ht.hash("dear") == 412
+        assert ht.hash("cat") == 312
+    `) 
+})
+```
+
+`HashTable.add("golf", "sport")` should add the key-value pair to the collection at key `424`.
+
+```js
+
+({ 
+    test: () => runPython(`
+    ht = HashTable()
+    ht.add("golf", "sport")
+    expected_value = {424: {'golf': 'sport'}}
+
+    assert ht.collection == expected_value
+    `) 
+})
+```
+
+`HashTable.add("dear", "friend")` and `HashTable.add("read", "book")` should add both the key-value pairs to the collection at index `412` as a nested dictionary.
+
+```js
+({ 
+    test: () => runPython(`
+ 
+    ht = HashTable()
+
+    ht.add("dear", "friend")
+    ht.add("read", "book")
+
+    expected_value = {
+        "dear": "friend",
+        "read": "book"
+    }
+
+    assert ht.collection.get(412) == expected_value
+    `) 
+})
+```
+
+When the `"golf", "sport"` key-value pair exists in the hash table, `HashTable.remove("golf")` should remove the key-value pair from the collection.
```

Comment:
```suggestion
When the `golf` key exists in the hash table, `HashTable.remove('golf')` should remove the `golf` key from the collection.
```
I wouldn't mention the value because it doesn't matter in this case. And in practice removing a key or removing a key-value pair means the same thing.